### PR TITLE
Add Control Port to Bedrock

### PR DESCRIPTION
**Pull request by tylerkaraszewski**
@cead22 @coleaeason @righdforsa 

[HOLD] for https://github.com/Expensify/Server-Expensify/pull/2137 to make auth tests compatible with this.

This adds the long-discussed but never implemented control port to Bedrock. Essentially, it adds a second command port, and a set of commands called "control commands". There are currently three control commands:
```
BeginBackup
SuppressCommandPort
ClearCommandPort
```

These are in addition to the existing signal-based handling for these actions, though the intention is to eventually remove signals as a mechanism for controlling bedrock and use the control port instead.

### What's better about using a control port than signals?

1. We can send more than a few commands, with names that make sense. Currently, to run a backup, we send `SIGHUP`. To close the command port, we send `SIGTTIN`. Nobody knows what `SIGTTIN` even exists for. Then we're almost out of useable signals unless we just start sending random numbers.
2. Control commands can have parameters. What if we want to send a timeout for how long to close the command port for? Easy with a control command, impossible with a signal.
3. Control commands can return results. Want to know if the command port actually opened or closed, or query some internal state? Easy with a control command, impossible with a signal.
4. Signals are messy, and we can avoiding dealing with [signal safety](http://man7.org/linux/man-pages/man7/signal-safety.7.html).

### Extra Bonus: Send Commands When the Command Port is Closed

Ever wanted to send a `Status` command to a bedrock node, but couldn't, because the command port was closed? Now you can. The control port is just an extra command port that always stays open, and only listens on localhost. You can send it any command you like. If the command port is closed though, you probably don't want to send regular auth commands.

### Notes

1. The command port defaults to port 9999, which was chosen pretty arbitrary.
2. Control commands can be sent to the regular command port as well, but will only be processed if they originated from a localhost address.

### Testing

Using `nc`, send the new control commands to bedrock and observe the results. They should mirror the existing behavior for signals.

### CODE REVIEW

**Comment by cead22** [comment]
Passing this review to @quinthar or @coleaeason since they're more familiar with signal handling than me (and I'll be off Thu and Fri)

**Comment by quinthar** [comment]
Do we have any security concerns around this?  Currently you need sudo
access to signal the Bedrock process, but this will let anybody on the box
do it.  (I'm assuming we only open this control port on localhost?)

**Comment by coleaeason** [comment]
Re: security concerns, anyone with access to the box also has sudo already. As long as this port is localhost only I don't see any security issues.

**Comment by tylerkaraszewski** [comment]
The port is only opened on localhost (by default, though if you pass in an address like 0.0.0.0, it will be opened on all interfaces), and it validates the remote IP of any socket before running a control command, enforcing that it's 127.0.0.1, so we should be good in that case. I've tested this locally, you can send control commands on either normal command port or the control port from localhost, and they work, but on the host machine (with bedrock running in the VM), you can't connect to the control port on the VM, and if you send a control command to the regular command port, it returns `401 Unauthorized`.

**Comment by coleaeason** [comment]
Removed the hold, I merged the other PR. @quinthar any more pushback on this? Otherwise can you  please review?

**Comment by quinthar** [comment]
Oh crap, I didn't realize this was waiting on my review.  Doing that now.

File: BedrockServer.cpp [hunk 1]
```diff
@@ -869,29 +869,44 @@ void BedrockServer::postPoll(fd_map& fdm, uint64_t& nextActivity) {
     if (!_suppressCommandPort && state == SQLiteNode::SLAVING && (masterVersion != _version)) {
         SINFO("Node " << _args["-nodeName"] << " slaving on version " << _version << ", master is version: "
               << masterVersion << ", not opening command port.");
-        suppressCommandPort(true);
+        suppressCommandPort("wrong master version", true);
```
Comment (quinthar) [inline BedrockServer.cpp]: This message is a bit confusing -- how about `master version mismatch`?  The master version is "right", just different.

File: BedrockServer.cpp [hunk 2]
```diff
@@ -869,29 +869,44 @@ void BedrockServer::postPoll(fd_map& fdm, uint64_t& nextActivity) {
...
     } else if (_suppressCommandPort && (state == SQLiteNode::MASTERING || (masterVersion == _version))) {
         // If we become master, or if master's version resumes matching ours, open the command port again.
         if (!_suppressCommandPortManualOverride) {
             // Only generate this logline if we haven't manually blocked this.
             SINFO("Node " << _args["-nodeName"] << " disabling previously suppressed command port after version check.");
         }
-        suppressCommandPort(false);
+        suppressCommandPort("master version valid", false);
```
Comment (quinthar) [inline BedrockServer.cpp]: Similarly, the master version isn't "valid", it's just the same.  So how about `master version match`

File: BedrockServer.cpp [hunk 3]
```diff
@@ -869,29 +869,44 @@ void BedrockServer::postPoll(fd_map& fdm, uint64_t& nextActivity) {
...
         _shutdownState.load() == RUNNING) {
         // Open the port
-        SINFO("Ready to process commands, opening command port on '" << _args["-serverHost"] << "'");
-        openPort(_args["-serverHost"]);
+        if (!_commandPort) {
+            SINFO("Ready to process commands, opening command port on '" << _args["-serverHost"] << "'");
+            _commandPort = openPort(_args["-serverHost"]);
+        }
+        if (!_controlPort) {
+            _controlPort = openPort(_args["-controlPort"]);
```
Comment (quinthar) [inline BedrockServer.cpp]: Can you add a matching `SINFO` explaining that we're also opening the control port?

File: BedrockServer.cpp [hunk 4]
```diff
@@ -1122,7 +1140,8 @@ void BedrockServer::_reply(BedrockCommand& command) {
     }
 }
 
-void BedrockServer::suppressCommandPort(bool suppress, bool manualOverride) {
+void BedrockServer::suppressCommandPort(const string& reason, bool suppress, bool manualOverride) {
+    SINFO((suppress ? "Suppressing" : "Clearing") << " command port due to: " << reason);
```
Comment (quinthar) [inline BedrockServer.cpp]: Can you put this line underneath the comment?  See [here](https://github.com/Expensify/Style-Guide/blob/master/php.md#blank-lines)

File: libstuff/STCPServer.cpp [hunk 5]
```diff
@@ -23,15 +23,25 @@ STCPServer::Port* STCPServer::openPort(const string& host) {
     return &*portIt;
 }
 
-void STCPServer::closePorts() {
+void STCPServer::closePorts(list<Port*> except) {
+    for (auto p : except) {
```
Comment (quinthar) [inline libstuff/STCPServer.cpp]: Can you either add a comment here and separate it from the next block by a blank line, or merge this into the next block by putting it under its comment?  (See [here](https://github.com/Expensify/Style-Guide/blob/master/php.md#blank-lines))

File: BedrockServer.cpp [hunk 6]
```diff
@@ -1304,3 +1348,25 @@ void BedrockServer::_postPollPlugins(fd_map& fdm, uint64_t nextActivity) {
         }
     }
 }
+
+void BedrockServer::_beginShutdown(const string& reason) {
+if (_shutdownState.load() == RUNNING) {
```
Comment (quinthar) [inline BedrockServer.cpp]: Can you indent this block?

File: main.cpp [hunk 7]
```diff
@@ -306,6 +307,10 @@ int main(int argc, char* argv[]) {
                 server.postPoll(fdm, nextActivity);
             }
             SINFO("Graceful bedrock shutdown complete");
+
```
Comment (quinthar) [inline main.cpp]: Please remove the extra line.   (See [here](https://github.com/Expensify/Style-Guide/blob/master/php.md#blank-lines))

**Pull request merged**